sjasm/z80.cpp

```diff
@@ -4512,6 +4544,161 @@ namespace Z80 {
+            }
+            else{
+                e[0] = 0xed; e[1] = 0x91;e[2] = GetByte(lp);
+                if(!comma(lp)){
+                    break;
+                }
+                e[3] = GetByte(lp);
+            }
+            EmitBytes(e);
+            /* (begin add) */
+            if (*lp && comma(lp)) {
+                continue;
+            } else {
+                break;
+            }
+        }while('o');
+    }
+
+    void OpCode_MIRROR() {
+        Z80Reg reg;
+        int e[3];
+        do{
+            e[0] = e[1] = e[2]= -1;
+            reg = GetRegister(lp);
+            switch(reg){
+                case Z80_A:
+                    e[0] = 0xed; e[1] = 0x24;
+                    break;
+                case Z80_DE:
+                    e[0] = 0xed; e[1] = 0x26;
+                    break;
+                default:
+                    break;
+            }
+            EmitBytes(e);
+            /* (begin add) */
+            if (*lp && comma(lp)) {
+                continue;
+            } else {
+                break;
+            }
+        }while('o');
+    }
+
+    void OpCode_TEST() {
+        int e[4];
+        do{
+            e[0] = 0xED;
+            e[1] = 0x27;
+            e[2] = GetByte(lp);
+            e[3] = -1;
+            EmitBytes(e);
+            /* (begin add) */
+            if (*lp && comma(lp)) {
+                continue;
+            } else {
+                break;
+            }
+        }while('o');
+    }
+
+	void OpCode_MUL() {
+		int e[3];
+		e[0] = e[1] = e[2] = -1;
+		if (GetRegister(lp)==Z80_D && comma(lp) && GetRegister(lp)==Z80_E){
+			e[0]=0xED;
+			e[1]=0x30;
+		}
+		EmitBytes(e);
+	}
+
+    //Swaps the high and low nibbles of the accumulator.
+    void OpCode_SWAPNIB() {
+        EmitByte(0xED);
+        EmitByte(0x23);
+    }
+
+    void OpCode_PIXELDN() {
+        EmitByte(0xED);
+        EmitByte(0x93);
+    }
+
+    void OpCode_PIXELAD() {
+        EmitByte(0xED);
+        EmitByte(0x94);
+    }
+
+    void OpCode_FILLDE() {
+        EmitByte(0xED);
+        EmitByte(0xB5);
+    }
+
+    void OpCode_LDIX() {
+        EmitByte(0xED);
+        EmitByte(0xA4);
+    }
+
+    void OpCode_LDWS() {
+        EmitByte(0xED);
+        EmitByte(0xA5);
+    }
+
+    void OpCode_LDIRX() {
+        EmitByte(0xED);
+        EmitByte(0xB4);
+    }
+
+    void OpCode_LDDX() {
+        EmitByte(0xED);
+        EmitByte(0xAC);
+    }
+
+    void OpCode_LDDRX() {
+        EmitByte(0xED);
+        EmitByte(0xBC);
+    }
+
+    void OpCode_LDIRSCALE() {
+        EmitByte(0xED);
+        EmitByte(0xB6);
+    }
+
+    void OpCode_LDPIRX() {
+        EmitByte(0xED);
+        EmitByte(0xB7);
+    }
+
+
+
+    void OpCode_OUTINB() {
+        EmitByte(0xED);
+        EmitByte(0x90);
+    }
+
+    void OpCode_POPX() {
+        EmitByte(0xED);
+        EmitByte(0x8B);
+    }
+
+    void OpCode_SETAE() {
+        EmitByte(0xED);
+        EmitByte(0x95);
+    }
+
```

Comment:
And this extension was done based on some older core 1.xx versions, as it's missing these instructions:
`push imm16` (BTW current `OpCode_PUSH` implementation can lead to buffer overflow - not part of this changeset, but needs fix)
`bsla | bsra | bsrl | bsrf | brlc de,b` and `jp (c)`

Reply:
I take back the buffer overflow comment about current OpCode_PUSH, it's actually protected... but I'm probably going to modify it in the similar way how other opcodes deal with multi-operands, as part of the Next fixes...